src/dma/mod.rs

```diff
@@ -743,8 +743,8 @@ macro_rules! db_transfer_def {
     };
 }
 
-db_transfer_def!(DBTransfer, init, StaticWriteBuffer, write_buffer, mut;);
-db_transfer_def!(ConstDBTransfer, init_const, StaticReadBuffer, read_buffer;
```

Comment:
Can you explain this change more?

My understanding is that `'static` is required because the hardware DMA transfer can continue regardless if the `Transfer` is forgotten. For example:

```rust
    let spi = spi.disable();
    let mut short_buffer = [0u8; 1024];
    let streams = StreamsTuple::new(dp.DMA1, ccdr.peripheral.DMA1);
    let config = DmaConfig::default().memory_increment(true);
    let mut transfer: Transfer<_, _, PeripheralToMemory, _, _> =
        Transfer::init(streams.0, spi, &mut short_buffer[..], None, config);
    
    transfer.start(|spi| {
        spi.enable_dma_tx();
        spi.inner_mut()
            .cr1
            .write(|w| w.ssi().slave_not_selected().spe().enabled());
        spi.inner_mut().cr1.modify(|_, w| w.cstart().started());
    });

    mem::forget(transfer); // <<<<<<<<

    loop {
        // &ref to buffer while DMA transfer is still writing to buffer
        info!("short_buffer = {:?}", short_buffer);
    }
```

This is undefined behavior because the compiler can assume that the memory doesn't change between subsequent reads of &-references. From a quick search I can't find anything documenting this explicitly, but it's [apparently an open question](https://github.com/rust-lang/unsafe-code-guidelines/issues/152) if it's even valid to use volatile reads on memory altered by external "threads" - which wouldn't be a question if it were valid for &-references.

Perhaps there could be an unsafe constructor for non-`'static` buffers which has the invariant that the buffer must live as long as the *hardware* DMA transfer?

Reply:
I forgot about this. I think an unsafe constructor is a good solution, I can't think of a safe way to do this.